File: androidApp/src/main/kotlin/com/riox432/civitdeck/ui/compare/ImageComparisonSlider.kt
package com.riox432.civitdeck.ui.compare

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.rememberTransformableState
import androidx.compose.foundation.gestures.transformable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import coil3.request.ImageRequest
import coil3.request.crossfade
import com.riox432.civitdeck.R
import com.riox432.civitdeck.ui.theme.Duration
import kotlin.math.roundToInt

/** Slider orientation for image comparison. */
enum class SliderOrientation { Horizontal, Vertical }

/**
 * A before/after image comparison slider with drag-to-reveal and pinch-to-zoom.
 *
 * The before image is clipped by the slider position, revealing the after
 * image underneath. Users can drag the divider to compare and pinch to zoom.
 */
@Composable
fun ImageComparisonSlider(
    beforeImageUrl: String,
    afterImageUrl: String,
    modifier: Modifier = Modifier,
    orientation: SliderOrientation = SliderOrientation.Horizontal,
) {
    var containerSize by remember { mutableStateOf(IntSize.Zero) }
    var sliderFraction by remember { mutableFloatStateOf(INITIAL_FRACTION) }
    var scale by remember { mutableFloatStateOf(MIN_SCALE) }
    var panOffset by remember { mutableStateOf(Offset.Zero) }

    val transformState = rememberTransformableState { zoomChange, panChange, _ ->
        val newScale = (scale * zoomChange).coerceIn(MIN_SCALE, MAX_SCALE)
        scale = newScale
        panOffset = if (newScale > MIN_SCALE) panOffset + panChange else Offset.Zero
    }

    Box(
        modifier = modifier
            .clipToBounds()
            .onSizeChanged { containerSize = it }
            .transformable(state = transformState),
    ) {
        val zoomModifier = Modifier.graphicsLayer {
            scaleX = scale
            scaleY = scale
            translationX = panOffset.x
            translationY = panOffset.y
        }

        // After image (bottom layer, fully visible)
        ComparisonImage(afterImageUrl, Modifier.fillMaxSize().then(zoomModifier))

        // Before image (top layer, clipped by slider)
        Box(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer {
                    clip = true
                    shape = SliderClipShape(sliderFraction, orientation)
                },
        ) {
            ComparisonImage(beforeImageUrl, Modifier.fillMaxSize().then(zoomModifier))
        }

        // Divider line
        SliderDivider(sliderFraction, orientation, containerSize)

        // Drag handle
        SliderHandle(sliderFraction, orientation, containerSize) { delta ->
            val max = when (orientation) {
                SliderOrientation.Horizontal -> containerSize.width.toFloat()
                SliderOrientation.Vertical -> containerSize.height.toFloat()
            }
            if (max > 0f) {
                val primary = when (orientation) {
                    SliderOrientation.Horizontal -> delta.x
                    SliderOrientation.Vertical -> delta.y
                }
                sliderFraction = (sliderFraction + primary / max).coerceIn(0f, 1f)
            }
        }
    }
}

// MARK: - Internal composables

@Composable
private fun ComparisonImage(imageUrl: String, modifier: Modifier = Modifier) {
    AsyncImage(
        model = ImageRequest.Builder(LocalContext.current)
            .data(imageUrl).crossfade(Duration.normal).build(),
        contentDescription = null,
        contentScale = ContentScale.Fit,
        modifier = modifier,
    )
}

@Composable
private fun SliderDivider(
    fraction: Float,
    orientation: SliderOrientation,
    size: IntSize,
) {
    Canvas(modifier = Modifier.fillMaxSize()) {
        when (orientation) {
            SliderOrientation.Horizontal -> {
                val x = size.width * fraction
                drawLine(Color.White, Offset(x, 0f), Offset(x, size.height.toFloat()), DIVIDER_PX)
            }
            SliderOrientation.Vertical -> {
                val y = size.height * fraction
                drawLine(Color.White, Offset(0f, y), Offset(size.width.toFloat(), y), DIVIDER_PX)
            }
        }
    }
}

@Composable
private fun SliderHandle(
    fraction: Float,
    orientation: SliderOrientation,
    containerSize: IntSize,
    onDrag: (Offset) -> Unit,
) {
    val halfHandle = HANDLE_DP / 2f
    val ox = when (orientation) {
        SliderOrientation.Horizontal -> containerSize.width * fraction - halfHandle
        SliderOrientation.Vertical -> containerSize.width / 2f - halfHandle
    }.roundToInt()
    val oy = when (orientation) {
        SliderOrientation.Horizontal -> containerSize.height / 2f - halfHandle
        SliderOrientation.Vertical -> containerSize.height * fraction - halfHandle
    }.roundToInt()

    Box(
        modifier = Modifier
            .offset { IntOffset(ox, oy) }
            .size(HANDLE_DP.dp)
            .clip(CircleShape)
            .background(Color.White.copy(alpha = HANDLE_ALPHA))
            .pointerInput(Unit) { detectDragGestures { _, d -> onDrag(d) } },
        contentAlignment = Alignment.Center,
    ) {
        val iconRes = when (orientation) {
            SliderOrientation.Horizontal -> R.drawable.ic_slider_horizontal
            SliderOrientation.Vertical -> R.drawable.ic_slider_vertical
        }
        Icon(
            painter = painterResource(id = iconRes),
            contentDescription = "Slider handle",
            tint = MaterialTheme.colorScheme.onSurface,
            modifier = Modifier.size(ICON_DP.dp),
        )
    }
}

private const val INITIAL_FRACTION = 0.5f
private const val MIN_SCALE = 1f
private const val MAX_SCALE = 5f
private const val DIVIDER_PX = 3f
private const val HANDLE_DP = 40f
private const val ICON_DP = 20f
private const val HANDLE_ALPHA = 0.85f

File: iosApp/iosApp/Features/Compare/ImageComparisonSlider.swift
import SwiftUI

/// Slider orientation for image comparison.
enum SliderOrientation {
    case horizontal
    case vertical
}

/// A before/after image comparison slider with drag-to-reveal and pinch-to-zoom.
///
/// The before image is clipped by the slider position, revealing the after
/// image underneath. Users can drag the divider to compare and pinch to zoom.
struct ImageComparisonSlider: View {
    let beforeImageUrl: String
    let afterImageUrl: String
    var orientation: SliderOrientation = .horizontal

    @State private var sliderFraction: CGFloat = initialFraction
    @State private var scale: CGFloat = 1.0
    @State private var offset: CGSize = .zero

    var body: some View {
        GeometryReader { geometry in
            let size = geometry.size

            ZStack {
                // After image (bottom layer, fully visible)
                comparisonImage(url: afterImageUrl, size: size)

                // Before image (top layer, clipped by slider)
                comparisonImage(url: beforeImageUrl, size: size)
                    .clipShape(SliderClipShape(fraction: sliderFraction, orientation: orientation))

                // Divider line
                dividerLine(size: size)

                // Drag handle
                sliderHandle(size: size)
            }
            .clipped()
            .gesture(magnificationGesture)
        }
    }

    // MARK: - Image

    private func comparisonImage(url: String, size: CGSize) -> some View {
        CachedAsyncImage(url: URL(string: url), maxPixelSize: imageMaxPixelSize) { phase in
            switch phase {
            case .success(let image):
                image
                    .resizable()
                    .scaledToFit()
                    .scaleEffect(scale)
                    .offset(offset)
            case .failure:
                Color.civitSurfaceVariant
                    .overlay {
                        Image(systemName: "photo")
                            .foregroundColor(.civitOnSurfaceVariant)
                            .accessibilityHidden(true)
                    }
            case .empty:
                Color.civitSurfaceVariant.shimmer()
            @unknown default:
                Color.civitSurfaceVariant
            }
        }
        .frame(width: size.width, height: size.height)
    }

    // MARK: - Divider

    private func dividerLine(size: CGSize) -> some View {
        Canvas { context, canvasSize in
            var path = Path()
            switch orientation {
            case .horizontal:
                let posX = canvasSize.width * sliderFraction
                path.move(to: CGPoint(x: posX, y: 0))
                path.addLine(to: CGPoint(x: posX, y: canvasSize.height))
            case .vertical:
                let posY = canvasSize.height * sliderFraction
                path.move(to: CGPoint(x: 0, y: posY))
                path.addLine(to: CGPoint(x: canvasSize.width, y: posY))
            }
            context.stroke(path, with: .color(.white), lineWidth: dividerWidth)
        }
        .allowsHitTesting(false)
    }

    // MARK: - Handle

    private func sliderHandle(size: CGSize) -> some View {
        let posX: CGFloat
        let posY: CGFloat

        switch orientation {
        case .horizontal:
            posX = size.width * sliderFraction
            posY = size.height / 2
        case .vertical:
            posX = size.width / 2
            posY = size.height * sliderFraction
        }

        return Circle()
            .fill(Color.civitSurface.opacity(handleAlpha))
            .frame(width: handleSize, height: handleSize)
            .overlay {
                Image(systemName: orientation == .horizontal
                      ? "arrow.left.and.right"
                      : "arrow.up.and.down")
                    .font(.civitIconSmallSemibold)
                    .foregroundColor(.civitOnSurface)
                    .accessibilityHidden(true)
            }
            .position(x: posX, y: posY)
            .gesture(dragGesture(size: size))
            .accessibilityLabel("Comparison slider")
            .accessibilityValue("\(Int(sliderFraction * 100))%")
    }

    // MARK: - Gestures

    private func dragGesture(size: CGSize) -> some Gesture {
        DragGesture()
            .onChanged { value in
                switch orientation {
                case .horizontal:
                    sliderFraction = (value.location.x / size.width).clamped(to: 0...1)
                case .vertical:
                    sliderFraction = (value.location.y / size.height).clamped(to: 0...1)
                }
            }
    }

    private var magnificationGesture: some Gesture {
        MagnificationGesture()
            .onChanged { value in
                scale = min(max(value, minScale), maxScale)
            }
            .onEnded { _ in
                withAnimation(MotionAnimation.fast) {
                    if scale < 1.2 {
                        scale = 1.0
                        offset = .zero
                    }
                }
            }
    }
}

// MARK: - Clip Shape

private struct SliderClipShape: Shape {
    let fraction: CGFloat
    let orientation: SliderOrientation

    func path(in rect: CGRect) -> Path {
        switch orientation {
        case .horizontal:
            return Path(CGRect(
                x: rect.minX, y: rect.minY,
                width: rect.width * fraction, height: rect.height
            ))
        case .vertical:
            return Path(CGRect(
                x: rect.minX, y: rect.minY,
                width: rect.width, height: rect.height * fraction
            ))
        }
    }
}

// MARK: - Constants

private let initialFraction: CGFloat = 0.5
private let minScale: CGFloat = 1.0
private let maxScale: CGFloat = 5.0
private let dividerWidth: CGFloat = 3
private let handleSize: CGFloat = 40
private let iconSize: CGFloat = 16
private let handleAlpha: Double = 0.85
private let imageMaxPixelSize: CGFloat = 1200

// MARK: - Comparable Clamping

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
